### ISSUE TITLE
Please add Support for Controlling Remote Schema Reading

### ISSUE THREAD

**Issue by keklabs** [zaggino/z-schema#107]
I already read the sample provided at https://github.com/zaggino/z-schema/blob/master/test/spec/AutomaticSchemaLoadingSpec.js
but the solution with repeated validation is not so effective as I need. 

As I understand the sample,  I have to

1. try validation
2. check for unresolved schema references and load them
3. repeat from 1) until all schema references are resolved

so when I will have deep (recursive) reference structure of schemas, than I must repeat and repeat validation, or  download schemas, parse them and search for internal references myself and after all run validation once. 

I have one additional requirement:  we need to specify references  by URIs, but our format of URI is independent on HTTP-URL format.  Our URI reprezents Address of Schema in some SchemaCatalog.  So I need to resolve ( download the referenced schema) by rutine. 

I would ask for support customizable Schema reading. The solution could be inspired by the same feature provided in Ruby library: https://github.com/ruby-json-schema/json-schema#controlling-remote-schema-reading

This solution could provide:
-  better and effective control for remote schema loading  - no repeated validations
-  could provide schema loading over different protokols, customized domain or from specialized schema catalog service, etc.

Sample usage:
```javascript
validator = new ZSchema();
validator.setSchemaReader(function(uri){
                            schemaBody = SchemaCatalog.loadSchema(uri); //custom loading implementation
                            return {uri:schemaBody}
                          });
validator.validate(jsonData,{"$ref":"urn:json-schema:0167-6423"});
 ```
Thanks so much for your help

**Comment by zaggino** [comment]
I can do this, but `validator.setSchemaReader(function(){})` would be only sync function so no promises, callback, async calls, reading from file system, etc.

**Comment by keklabs** [comment]
In my opinion it should be OK, it is synchronous feature, the validator should not continue with validation until the reference is resolved, because the remote schema is needed for validation of current element. 
 
The resolved schema will be cached and reused for other validations, so next validate will be faster, when no other unknown reference will be found.

**Comment by zaggino** [comment]
Implemented in `3.10.0`,  see https://github.com/zaggino/z-schema#remote-references-and-schemas and https://github.com/zaggino/z-schema/blob/5c77227439a65214db9e2dabad04198357332700/test/ZSchemaTestSuite/Issue107.js
